### PR TITLE
Raft + Transactions Support

### PR DESCRIPTION
**Pull request by absolutelightning**
* Removed all the TODO in https://github.com/absolutelightning/treds/pull/4
* Add Transaction Support

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -102,6 +108,12 @@ func (ts *Server) OnBoot(_ gnet.Engine) gnet.Action {
 			time.Sleep(100 * time.Millisecond)
 		}
 	}()
+	go func() {
+		for {
+			time.Sleep(60 * time.Second)
+			ts.raft.Snapshot()
+		}
+	}()
```
Comment (dhiaayachi) [inline server/server.go:116]: You don't need to run snapshots periodically, raft will do it by it self to compact logs. See [here](https://github.com/hashicorp/raft/blob/2b0032e390b888b0058d221989886d090b9d9609/api.go#L615-L615)

You can implement a snapshot command though that would take a user snapshot and backup the DB.

File: server/treds_fsm.go [hunk 2]
```diff
@@ -26,17 +29,50 @@ func (t TredsFsm) Apply(log *raft.Log) interface{} {
...
+	data, err := io.ReadAll(old)
+	if err != nil {
+		return err
+	}
+	ts := store.NewTredsStore()
+	err = ts.Restore(data)
+	if err != nil {
+		return err
+	}
+	t.tredsStore = ts
```
Comment (dhiaayachi) [inline server/treds_fsm.go:74]: I'm not sure where `Restore` could be called, but this is not thread safe. You need to make the `tredsStore` pointer atomic to make it safe.

**Review by dhiaayachi — CHANGES_REQUESTED** [review]
@absolutelightning I added few comments inline. I think we should start considering some testing at this point 😅 but maybe in a separate PR.

File: server/server.go (hunk 1 above)
Comment (absolutelightning) [reply server/server.go:116]: Oh got it.

Comment (absolutelightning) [reply server/server.go]: I see that function is there, but when I run the server - I don't see snapshot creating.

File: server/treds_fsm.go (hunk 2 above)
Comment (absolutelightning) [reply server/treds_fsm.go]: @dhiaayachi the system is running a single thread, is this required?

**Pull request merged**